runtime/runtime/src/congestion_control.rs

```diff
@@ -0,0 +1,372 @@
+) -> Result<Gas, IntegerOverflowError> {
+    match &receipt.receipt {
+        ReceiptEnum::Action(action_receipt) => {
+            // account for gas guaranteed to be used for executing the receipts
+            let prepaid_exec_gas = safe_add_gas(
+                total_prepaid_exec_fees(config, &action_receipt.actions, &receipt.receiver_id)?,
+                config.fees.fee(ActionCosts::new_action_receipt).exec_fee(),
+            )?;
+            // account for gas guaranteed to be used for creating new receipts
+            let prepaid_send_gas = total_prepaid_send_fees(config, &action_receipt.actions)?;
+            let prepaid_gas = safe_add_gas(prepaid_exec_gas, prepaid_send_gas)?;
+
+            // account for gas potentially used for dynamic execution
+            let gas_attached_to_fns = total_prepaid_gas(&action_receipt.actions)?;
+            let gas = safe_add_gas(gas_attached_to_fns, prepaid_gas)?;
+
+            Ok(gas)
+        }
+        ReceiptEnum::Data(_data_receipt) => {
+            // Data receipts themselves don't cost gas to execute, their cost is
+            // burnt at creation. What we should count, is the gas of the
+            // postponed action receipt. But looking that up would require
+            // reading the postponed receipt from the trie.
+            // Thus, the congestion control MVP does not account for data
+            // receipts or postponed receipts.
+            Ok(0)
+        }
+        ReceiptEnum::PromiseYield(_) => {
+            // The congestion control MVP does not account for yielding a promise.
+            // It does not burn any gas, so it's unclear what we could account for.
+            Ok(0)
+        }
+        ReceiptEnum::PromiseResume(_) => {
+            // The congestion control MVP does not account for resuming a promise.
+            // Similar to a data receipt, it would be difficult to find the cost
+            // of it without expensive state lookups.
+            Ok(0)
+        }
+    }
+}
+
+/// Iterate all columns in the trie holding unprocessed receipts and
+/// computes the storage consumption as well as attached gas.
+///
+/// This is an IO intensive operation! Only do it to bootstrap the
+/// `CongestionInfo`. In normal operation, this information is kept up
+/// to date and passed from chunk to chunk through chunk extra fields.
+pub fn compute_congestion_info(
+    trie: &dyn near_store::TrieAccess,
+    config: &RuntimeConfig,
+    shard_id: ShardId,
+    other_shard_ids: &[ShardId],
+    congestion_seed: u64,
+) -> Result<CongestionInfo, StorageError> {
+    let mut receipt_bytes: u64 = 0;
+    let mut delayed_receipts_gas: u128 = 0;
+    let mut buffered_receipts_gas: u128 = 0;
+
+    let delayed_receipt_queue = &DelayedReceiptQueue::load(trie)?;
+    for receipt_result in delayed_receipt_queue.iter(trie) {
+        let receipt = receipt_result?;
+        let gas = receipt_congestion_gas(&receipt, config).map_err(int_overflow_to_storage_err)?;
```

Comment:
Can you replace all of those .map_err(int_overflow_to_storage_err) with a From trait implementation for the errors?

Reply:
Hm, are you sure about that? Implementing it would need to be done globally, which means the conversion can happen everyhwere. But I don't think `IntegerOverflowError` is a `StorageError` makes sense in all places.

Reply:
It is in general a bad practice to have an implementation like that. It, for one, fails to properly establish the necessary context to determine which of the places the error originates from. 

I would argue that this `int_overflow_to_storage_err` fails to do as well. I think the error chain returned in this particular location should read something like:

* failed to compute receipt congestion gas
  * could not add that particular type of gas
    * integer overflow

but us not doing so is a problem shared by the overall codebase, unfortunately, so I'm not going to insist here.

Reply:
I was recently debugging some tests and suffered from this lack of context. I ended up replacing `?` with `unwrap()`s until I narrowed it down to where the issue was actually happening. How would you go about adding this context here? I made an attempt to do so for one particular error in #11172 but I'm curious if there is a better or more rusty way to do it.

Reply:
`map_err` is the way to do it, but it needs to be consistently applied everywhere, and not just selectively where types need to be made to match. [I have a blog post on this](https://kazlauskas.me/entries/errors), but a summary of it is that when theres a `fallible_operation()`, it has to be followed immediately by `map_err(Error::ThisSpecificFallibleOperation)` where the error uniquely describes this `fallible_operation` semantically.